### ISSUE TITLE
IndexOutOfBoundsException when extracting lucene-core-6.1.0.jar > Lucene54DocValuesConsumer.class

### ISSUE THREAD

**Issue by OlivierJaquemet** [jeantessier/dependency-finder#2]
The following exceptions occurs when trying to extract `lucene-core-6.1.0.jar` .

To reproduce : 
- Download `lucene-6.1.0.zip` from http://apache.crihan.fr/dist/lucene/java/6.1.0/ 
- Unzip jar `lucene-6.1.0\core\lucene-core-6.1.0.jar` in the archive
- Start dependency extraction on this jar

Tested with both DependencyFinder-1.2.1-beta4 build and current master branch as of today.

```
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index: 2925, Size: 78
        at java.util.ArrayList.rangeCheck(ArrayList.java:635)
        at java.util.ArrayList.get(ArrayList.java:411)
        at com.jeantessier.classreader.impl.Classfile.getRawClass(Classfile.java:200)
        at com.jeantessier.classreader.impl.Classfile.getClassName(Classfile.java:204)
        at com.jeantessier.classreader.impl.Classfile.<init>(Classfile.java:104)
        at com.jeantessier.classreader.impl.Classfile.<init>(Classfile.java:74)
        at com.jeantessier.classreader.impl.DefaultClassfileFactory.create(DefaultClassfileFactory.java:41)
        at com.jeantessier.classreader.impl.DefaultClassfileFactory.create(DefaultClassfileFactory.java:39)
        at com.jeantessier.classreader.TransientClassfileLoader.load(TransientClassfileLoader.java:70)
        at com.jeantessier.classreader.ClassfileLoaderEventSource.load(ClassfileLoaderEventSource.java:133)
        at com.jeantessier.classreader.ZipClassfileLoader.load(ZipClassfileLoader.java:127)
        at com.jeantessier.classreader.JarClassfileLoader.load(JarClassfileLoader.java:53)
        at com.jeantessier.classreader.ClassfileLoaderEventSource.load(ClassfileLoaderEventSource.java:93)
        at com.jeantessier.classreader.ClassfileLoader.load(ClassfileLoader.java:55)
        at com.jeantessier.dependencyfinder.cli.DependencyExtractor.doProcessing(DependencyExtractor.java:74)
        at com.jeantessier.dependencyfinder.cli.Command.process(Command.java:187)
        at com.jeantessier.dependencyfinder.cli.Command.run(Command.java:109)
        at com.jeantessier.dependencyfinder.cli.DependencyExtractor.main(DependencyExtractor.java:101)
```

The culprit class appears to be Lucene54DocValuesConsumer.class whose source is here : 
https://github.com/apache/lucene-solr/blob/releases/lucene-solr/6.1.0/lucene/core/src/java/org/apache/lucene/codecs/lucene54/Lucene54DocValuesConsumer.java

Not being very familiar with the project, I could not quickly find an explanation nor a fix.  

I suspect an issue with the lucene jar / class / source because Java Decompiler 1.4 ( https://github.com/java-decompiler/jd-gui) also fails to open the same class with a `// INTERNAL ERROR //` message.
I have also entered an issue on the JD project : 
https://github.com/java-decompiler/jd-gui/issues/140

Could it be related to this code and the corresponding byte code ?
https://github.com/apache/lucene-solr/blob/releases/lucene-solr/6.1.0/lucene/core/src/java/org/apache/lucene/codecs/lucene54/Lucene54DocValuesConsumer.java#L703

**Comment by OlivierJaquemet** [comment]
Same error occurs for class `org\apache\lucene\util\AttributeImpl.class` in same `lucene-core-6.1.0.jar` 

Corresponding source file : 
https://github.com/apache/lucene-solr/blob/releases/lucene-solr/6.1.0/lucene/core/src/java/org/apache/lucene/util/AttributeImpl.java

If thoses 2 classes are remove from jar, extraction can be completed successfully.

**Comment by jeantessier** [comment]
Thanks for reporting this.  I will look into it and let you know what I find.

**Comment by OlivierJaquemet** [comment]
@jeantessier thanks for looking into it, and most of all, kudos to you for creating Dependency Finder, absolutely awesome tool :)

You may want to check the discussion in the corresponding java-decompiler issue as they will certainly face the same technical challenge. Either you'll be the first to share your findings, or else you'll benefit from theirs.

**Comment by jeantessier** [comment]
Here's what I've found so far.  I used `ClassReader` on each `.class` file in `lucene-core-6.1.0.jar` to figure out which ones were giving out parsing errors.  I found three:

* `org/apache/lucene/codecs/lucene54/Lucene54DocValuesConsumer$1.class`
* `org/apache/lucene/codecs/lucene54/Lucene54DocValuesConsumer$2.class`
* `org/apache/lucene/util/AttributeImpl.class`

I turned on super-verbose mode by tweaking `log4j.properties` in my Dependency Finder installation to add `log4j.logger.com.jeantessier.classreader=DEBUG` and I ran `ClassReader` on `AttributeImpl.class` again:

    ClassReader \
        -verbose AttributeImpl.out.txt \
        org/apache/lucene/util/AttributeImpl.class

The output shows the major version is 52, which means the class was compiled J2SE 8 (see [this stackoverflow](http://stackoverflow.com/questions/22489398/unsupported-major-minor-version-52-0)).

I have not looked at changes to the classfile format in Java 8.  After reading the minor and major version numbers from the classfile, the next thing is supposed to be the constant pool.  The verbose output from `ClassReader` shows a number of "Unknown Tag" as it's trying to read the constant pool, so I'm thinking there might have been a change in the classfile format that is tripping my parser.

Next step will be to research changes to the classfile format, but that'll have to wait until tonight.

**Comment by jeantessier** [comment]
Looking at the [JVM Spec for Java 7](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.4.10), it looks that there are at least 3 new types of entries in the constant pool:

* `MethodHandle` (15)
* `MethodType` (16)
* `InvokeDynamic` (18)

When trying to parse `AttributeImpl.class`, the first unknown tag was 18, an `InvokeDynamic` entry.

I would need to augment my parser to take them into account, plus whatever structures they might point to.  Even if only so I can get through the decompilation.  Shouldn't be too hard.

**Comment by jeantessier** [comment]
The good news is that the [JVM Spec for Java 8](https://docs.oracle.com/javase/specs/jvms/se8/html/index.html) does not add any new constant pool entries.  I'll have to do a bit of research to see what's different there too.

**Comment by jeantessier** [comment]
I checked in some code to `master` to parse new missing constant pool entries, so at least the rest of Dependency Finder can get to work.

**Comment by OlivierJaquemet** [comment]
Thanks a lot Jean !

**Comment by jeantessier** [comment]
Included in release 1.2.1-beta5.
